Publish Arm's internal documentation

michalpasztamobica opened the pull request:
This is the first part of the exercise aiming at establishing the approach, files locations, agreeing on using `rst` format, etc.

The highlights of this commits are:
1) New documents added in `.rst` format in the main `docs/` directory. Being aware that most documentation so far was only available in the `.nroff` format we see some befits of choosing `rst` for new docs.
* `rst` to html on the fly, so all our documents will be viewable online
* `rst` files can also be used for the wiki (unless we decide to move wiki to main repository)
* current build tools allow us to convert `rst` to html easily using Sphinx and generate neat html files (see [html_docs_flang.zip](https://github.com/flang-compiler/flang/files/5401472/html_docs_flang.zip)), which we could set up somewhere - this is why I would prefer `rst` over `md` format.
2) `docs/ReleaseNotes.rst` is removed - it has never been used
3) Glossary is introduced based on ARM's glossary with some more phrases added.

Please review the code thinking wide - if this structure is approved we'll add more articles and documents in the near future.

@kiranchandramohan , @RichBarton-Arm , thanks for internal reviews and sharing Arm's internal documentation!

kiranchandramohan commented:
ping. Is this OK with everyone? This is only a documentation change.

bryanpkc commented on tools/flang1/utils/n2rst/CMakeLists.txt:
```diff
@@ -16,8 +16,6 @@ target_include_directories(fen2rst
 # Convert nroff documentation to Sphinx format.
 
-file(MAKE_DIRECTORY ${FLANG1_DOC_BIN_DIR})
-
```
Don't we need to create the directory before writing files to it?

bryanpkc replied:
`tools/flang1/flang1exe/CMakeLists.txt` also creates the directory, but only when `FLANG_INCLUDE_DOCS`  is true. That may result in situation where the `n2rst` build needs the directory but cannot find it.

michalpasztamobica replied:
This directory is created by two other `CMakeLists.txt`s already, but I guess you're right - it doesn't hurt if we keep this command in here, as it generates no errors or warnings and will keep this script working if other `CMakeLists.txt`s alter.

bryanpkc replied:
Thanks.

gklimowicz reviewed (approved):
These new documents got built for me on OpenPower.

bryanpkc reviewed (approved):
LGTM.

michalpasztamobica commented:
The reviews were still pending, so I allowed myself to expand this PR instead of creating a new one, as I already got green light from @RichBarton-Arm  to publish further docs. I hope this is OK?
@gklimowicz , @bryanpkc , would you please look at the other commit I just pushed and review that part?
@shivaramaarao , @RichBarton-Arm, your reviews will be very welcome for the whole PR.
@kiranchandramohan , once the approvals are there, I'll squash the two commits into one.

gklimowicz commented:
> The reviews were still pending, so I allowed myself to expand this PR instead of creating a new one, as I already got green light from @RichBarton-Arm to publish further docs. I hope this is OK?

Some of us had already reviewed it. In my case, I would have preferred to have a new PR for the additional changes.

gklimowicz reviewed (approved):
LGTM

bryanpkc commented on docs/glossary.rst:
```diff
@@ -9,11 +9,13 @@ Flang glossary
 | AST  | Abstract Syntax Tree.                                                                      |
 +------+--------------------------------------------------------------------------------------------+
-| BE   | (used heavily in code and comments, short for "backend"                                    |
+| BE   | (used heavily in code and comments), short for "backend"                                   |
```
Nit: There is no need for a comma after a parenthesized clause, and I would put the main definition first, e.g.
```
short for "backend" (used heavily in code and comments)
```

bryanpkc replied:
The same comment applies to all rows below that have parenthetical clauses.

bryanpkc commented on docs/debugging_flang.rst:
```diff
@@ -0,0 +1,214 @@
++--------------+--------------------------------------------------------------------------------------------+
+| Option       | Explanation                                                                                |
++==============+============================================================================================+
+| filename     | source file name (method 1 of 2)                                                           |
```
Would you consider styling all options in this table consistently as preformatted code? For example <code>\`\`filename\`\`</code>, <code>\`\`opt\`\`</code>, <code>\`\`x\`\`</code>, <code>\`\`qq\`\`</code>, etc.

bryanpkc commented on docs/debugging_flang.rst:
```diff
@@ -0,0 +1,214 @@
++--------------+--------------------------------------------------------------------------------------------+
+| x            | magic 'x' parameter                                                                        |
++--------------+--------------------------------------------------------------------------------------------+
+| y            | magic 'y' parameter                                                                        |
```
Should this explain that `y` flags turn off the corresponding `x` flags?

michalpasztamobica commented:
I am sorry @gklimowicz , next time in a situation like this I'll set up a new PR. Thanks for a quick review!
@bryanpkc , I hope I addressed all your remarks? I put them in a separate commit, which I will squash once I get all reviews.

@shivaramaarao , @RichBarton-Arm , any remarks to this PR?

bryanpkc reviewed (approved):
Thanks, LGTM.

shivaramaarao reviewed (approved):
Looks fine for me.

RichBarton-Arm commented:
Thanks @michalpasztamobica

bryanpkc commented:
@RichBarton-Arm Please rebase and merge; do not create merge commit in the git history.

RichBarton-Arm commented:
@bryanpkc - arg, sorry. I should have left it to Kiran to push the button.

Shall we try to fix it up?

bryanpkc commented:
@kiranchandramohan @gklimowicz @shivaramaarao What do you think? It's probably okay to fix the branch right?

Is there a setting in GitHub to prevent non-fast-forward merges?

michalpasztamobica commented:
Please, feel free to squash the three commits into one, if editing the master branch.

kiranchandramohan commented:
I am OK with fixing. We can discuss it tomorrow during the classic flang call.
